routes/users.router.js

```diff
@@ -0,0 +1,89 @@
+const express = require("express");
+const router = express.Router();
+const {User} = require("../models/user.model");
+const {extend} = require("lodash");
+const e = require("express");
+
+router.route("/")
+.post( async (req,res)=>{
+    try {
+        const userData = req.body;
+        
+        const user = await User.findOne({username: userData.username});
+
+        if(user){
+            res.status(409).json({success: false, message: "Account already exists for this email, please reset password if forgotten"});
+            return;
+        }
+
+        const NewUser = new User(userData);
+        const addedUserDataFromDb = await NewUser.save();
+
+        res.status(201).json({response: {firstname: addedUserDataFromDb.firstname, userId: addedUserDataFromDb._id}, success: true})
+
+    }  catch(error){
+        res.status(500).json({success:false, message: "Request failed please check errorMessage key for more details", errorMessage: error.message })
+    }
+})
+
+router.route("/authenticate")
+.post( async(req,res)=>{
+    try {
+        const username = req.get("username");
+        const password = req.get("password")
+        const user = await User.findOne({username});
+        
+        if(!user){
+            res.status(401).json({success:false, message: "Username is incorrect!"});
+            return;
+        }else if(user.password === password){
+            res.status(200).json({response: {firstname: user.firstname, userId: user._id}, success: true});
+            return;
+        }
+        res.status(401).json({ message : "Password is incorrect!", success : false });
+        
+    }  catch(error){
+        
+        res.status(500).json({success:false, message: "Request failed please check errorMessage key for more details", errorMessage: error.message })
+    }
+})
+
+router.param("username", async(req, res, next, id)=>{
+    const user = await User.findOne({username: id});
+        
+        if(!user){
+            res.status(404).json({success:false, message: "Username does not exist!"});
+            return;
+        }
+        req.user = user;
+        next();
+})
+
+router.route("/:username")
+.get( async(req,res)=>{
+    try {
+        const { user } = req;
+        res.status(200).json({ response :  {username: user.username, firstname: user.firstname, lastname: user.lastname, userId: user._id}, success : true })
```

Comment:
Instead of writing user. user. many times, we can make the user.password as null/ undefined and send it back entire object right?

Reply:
It's a great idea, I will do that